### DISCUSSION TITLE
Sending players through the velocity server to other servers

### DISCUSSION THREAD

**Discussion by DevBlokJochem** [Minestom/Minestom#1414] [Help] [answered]
Hi, I have a lobby and a game server. I want to send a player from the lobby to the game server. How can I create this in minestom?

**Comment by FluxCapacitor2** [comment, marked as answer]
Velocity supports the BungeeCord plugin messaging channel, so you can send a plugin message to a player using a [`PluginMessagePacket`](https://javadoc.minestom.net/net/minestom/server/network/packet/server/play/PluginMessagePacket.html), it will be intercepted by Velocity (if you have the option enabled in the Velocity [config](https://velocitypowered.com/wiki/users/configuration/))

More info can be found on the Spigot website: https://www.spigotmc.org/wiki/bukkit-bungee-plugin-messaging-channel/
You are probably looking for the [`Connect`](https://www.spigotmc.org/wiki/bukkit-bungee-plugin-messaging-channel/#connect) message.

Here's a short code example:
```java
ByteArrayDataOutput out = ByteStreams.newDataOutput();
out.writeUTF("Connect");
out.writeUTF("servername");
player.sendPacket(new PluginMessagePacket("BungeeCord", out.toByteArray());
```

**Reply by dev-hydrogen** [reply]
What about if theres no players online? You cant rely on the fact that youre always able to send a plugin message to player. This may also be a security risk *if* the player ends up receiving the packet (if the packet has anything "secret" only the servers should know in it)

**Reply by Moulberry** [reply]
If there are no players online, who are you trying to send to a different server?
None of the bungee packets contain any secrets, either.

**Reply by dev-hydrogen** [reply]
> If there are no players online, who are you trying to send to a different server? None of the bungee packets contain any secrets, either.

I was talking more generally about the bungeecord messaging channel, especially the Forward channel. Though theoretically an empty server could want to request to send a specific player or all players to itself?

**Reply by tjalp** [reply]
In that case you're probably better off just using a socket, message broker (Redis, RabbitMQ) or something alike.

**Reply by muelleel** [reply]
I could not resolve the class `ByteArrayDataOutput`, and other similar classes like the `BinaryWriter` or `ByteArrayOutputStream` didn't work.
To make it work you can add the following dependency (gradle in my case)
`implementation 'com.google.guava:guava:31.0.1-jre'`

after that you can use the `ByteArrayDataOutput` like in the example.

Also note that you may have to use `bungeecord:main` instead of `BungeeCord` as the channel.

```java
    private static final String bungeeTargetSelector = "bungeecord:main";
    public static void connect(Player p, String bungeeServerTargetName) {
        ByteArrayDataOutput out = ByteStreams.newDataOutput();
        out.writeUTF("Connect");
        out.writeUTF(bungeeServerTargetName);
        p.sendPacket(new PluginMessagePacket(bungeeTargetSelector, out.toByteArray()));
    }
```
